Libraries/LibWeb/HTML/Parser/HTMLParser.cpp

```diff
@@ -4351,103 +4355,143 @@ void HTMLParser::process_using_the_rules_for_foreign_content(HTMLToken& token)
 // https://html.spec.whatwg.org/multipage/parsing.html#reset-the-insertion-mode-appropriately
 void HTMLParser::reset_the_insertion_mode_appropriately()
 {
+    // 1. Let last be false.
+    bool last = false;
+
+    // 2. Let node be the last node in the stack of open elements.
     for (ssize_t i = m_stack_of_open_elements.elements().size() - 1; i >= 0; --i) {
-        bool last = i == 0;
-        // NOTE: When parsing fragments, we substitute the context element for the root of the stack of open elements.
+        // 3. Loop: If node is the first node in the stack of open elements, then set last to true, and, if the parser
+        //    was created as part of the HTML fragment parsing algorithm (fragment case), set node to the context
+        //    element passed to that algorithm.
         GC::Ptr<DOM::Element> node;
+        if (i == 0)
+            last = true;
+
         if (last && m_parsing_fragment) {
             node = m_context_element.ptr();
         } else {
             node = m_stack_of_open_elements.elements().at(i).ptr();
         }
 
-        if (node->namespace_uri() != Namespace::HTML)
-            continue;
-
-        if (node->local_name() == HTML::TagNames::select) {
-            if (!last) {
-                for (ssize_t j = i; j > 0; --j) {
-                    auto& ancestor = m_stack_of_open_elements.elements().at(j - 1);
-
-                    if (is<HTMLTemplateElement>(*ancestor))
-                        break;
-
-                    if (is<HTMLTableElement>(*ancestor)) {
-                        m_insertion_mode = InsertionMode::InSelectInTable;
-                        return;
+        // NOTE: The following steps only apply to HTML elements, so skip ones that aren't.
+        if (node->namespace_uri() == Namespace::HTML) {
+            // 4. If node is a select element, run these substeps:
+            if (node->local_name() == HTML::TagNames::select) {
```

Comment:
Is there any difference between this and `is<HTMLSelectElement>(*node)`?

Reply:
Conceptually, I don't think so? In practice, we have no `fast_is<>` for that so it'd be slower.